File: penpals-frontend/src/services/account.ts
// Account management service

import { ApiClient } from './api';
import { User, Classroom } from './auth';

export interface AccountDetails {
  account: User & { classroom_count: number };
  classrooms: Classroom[];
}

export interface AccountStats {
  account_id: number;
  total_classrooms: number;
  total_connections: number;
  unique_interests: number;
  account_created: string;
}

export interface UpdateAccountData {
  email?: string;
  password?: string;
  organization?: string;
}

const getAccountWithClassroomCount = (account: any) => {
  if (!account) {
    return account;
  }

  return {
    ...account,
    classroom_count:
      account.classroom_count ?? account.profile_count ?? account.total_profiles ?? account.total_classrooms,
  };
};

const getClassroomsFromResponse = (response: any): Classroom[] =>
  response.classrooms ?? response.profiles ?? [];

/**
 * Account management service
 */
export class AccountService {
  /**
   * Get current account details with all classrooms
   */
  static async getAccountDetails(): Promise<AccountDetails> {
    const response = await ApiClient.get<any>('/account');
    return {
      account: getAccountWithClassroomCount(response.account ?? response.user ?? {}),
      classrooms: getClassroomsFromResponse(response),
    };
  }

  /**
   * Update account information
   */
  static async updateAccount(data: UpdateAccountData): Promise<{
    msg: string;
    account: User;
  }> {
    return ApiClient.put('/account', data);
  }

  /**
   * Delete current account and all associated classrooms
   */
  static async deleteAccount(): Promise<{
    msg: string;
    deleted_classrooms: number;
  }> {
    const response = await ApiClient.delete('/account');
    
    // Clear token after successful deletion
    ApiClient.clearToken();
    
    return response;
  }

  /**
   * Get all classrooms for the current account
   */
  static async getAccountClassrooms(): Promise<{
    classrooms: (Classroom & { friends_count: number })[];
    total_count: number;
    account_id: number;
  }> {
    const response = await ApiClient.get<any>('/account/profiles');
    return {
      classrooms: getClassroomsFromResponse(response),
      total_count:
        response.total_count ?? response.total_profiles ?? response.total_classrooms ?? response.count ?? 0,
      account_id: response.account_id ?? response.profile_account_id ?? 0,
    };
  }

  /**
   * Get account statistics
   */
  static async getAccountStats(): Promise<AccountStats> {
    const response = await ApiClient.get<any>('/account/stats');
    return {
      account_id: response.account_id,
      total_classrooms: response.total_classrooms ?? response.total_profiles ?? 0,
      total_connections: response.total_connections ?? response.total_profile_connections ?? 0,
      unique_interests: response.unique_interests ?? response.unique_profile_interests ?? 0,
      account_created: response.account_created ?? response.created_at,
    };
  }

  /**
   * Validate account update data
   */
  static validateUpdateData(data: UpdateAccountData): {
    isValid: boolean;
    errors: string[];
  } {
    const errors: string[] = [];

    if (data.email !== undefined) {
      if (!data.email.trim()) {
        errors.push('Email cannot be empty');
      } else if (!/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/.test(data.email)) {
        errors.push('Invalid email format');
      }
    }

    if (data.organization !== undefined && data.organization.length > 120) {
      errors.push('Organization name too long (max 120 characters)');
    }

    if (data.password !== undefined) {
      if (data.password.length < 8) {
        errors.push('Password must be at least 8 characters long');
      }
      
      if (!/[A-Z]/.test(data.password)) {
        errors.push('Password must contain at least one uppercase letter');
      }
      
      if (!/[a-z]/.test(data.password)) {
        errors.push('Password must contain at least one lowercase letter');
      }
      
      if (!/\d/.test(data.password)) {
        errors.push('Password must contain at least one digit');
      }
      
      if (!/[^A-Za-z0-9]/.test(data.password)) {
        errors.push('Password must contain at least one special character');
      }
    }

    return {
      isValid: errors.length === 0,
      errors,
    };
  }
}

File: penpals-backend/src/account.py
"""
Account management endpoints for teacher accounts.
Handles account CRUD operations, password updates, and multi-classroom management.
"""

from flask import Blueprint, jsonify, request
from flask_jwt_extended import jwt_required, get_jwt_identity
from werkzeug.security import generate_password_hash
from models import db, Account, Profile

account_bp = Blueprint('account', __name__)

@account_bp.route('/api/account', methods=['GET'])
@jwt_required()
def get_account():
    """Get current account details with all classrooms"""
    account_id = get_jwt_identity()
    account = Account.query.get(account_id)
    
    if not account:
        return jsonify({"msg": "Account not found"}), 404
    
    classrooms = []
    for classroom in account.classrooms:
        classrooms.append({
            "id": classroom.id,
            "name": classroom.name,
            "location": classroom.location,
            "latitude": classroom.lattitude,
            "longitude": classroom.longitude,
            "class_size": classroom.class_size,
            "availability": classroom.availability,
            "interests": classroom.interests
        })
    
    return jsonify({
        "account": {
            "id": account.id,
            "email": account.email,
            "organization": account.organization,
            "created_at": account.created_at.isoformat()
        },
        "classrooms": classrooms
    }), 200


@account_bp.route('/api/account', methods=['PUT'])
@jwt_required()
def update_account():
    """Update account information"""
    account_id = get_jwt_identity()
    account = Account.query.get(account_id)
    
    if not account:
        return jsonify({"msg": "Account not found"}), 404
    
    data = request.json
    
    if 'email' in data:
        # Check if email is already taken by another account
        existing = Account.query.filter_by(email=data['email']).first()
        if existing and existing.id != account.id:
            return jsonify({"msg": "Email already in use"}), 409
        account.email = data['email']
    
    if 'organization' in data:
        account.organization = data['organization']
    
    if 'password' in data:
        password = data['password']
        capital_letters = [chr(i) for i in range(ord('A'), ord('Z')+1)]
        lowercase_letters = [chr(i) for i in range(ord('a'), ord('z')+1)]
        digits = [str(i) for i in range(10)]
        
        has_upper = any(c in capital_letters for c in password)
        has_lower = any(c in lowercase_letters for c in password)
        has_digit = any(c in digits for c in password)
        has_special = any(
            c not in capital_letters and c not in lowercase_letters and c not in digits
            for c in password
        )
        
        if not (len(password) >= 8 and has_upper and has_lower and has_digit and has_special):
            return jsonify({
                "msg": "Password must be at least 8 characters and include one uppercase, one lowercase, one digit, and one special character."
            }), 400
        
        account.password_hash = generate_password_hash(password)
    
    db.session.commit()
    
    return jsonify({"msg": "Account updated successfully"}), 200


@account_bp.route('/api/account', methods=['DELETE'])
@jwt_required()
def delete_account():
    """Delete account and all associated classrooms"""
    account_id = get_jwt_identity()
    account = Account.query.get(account_id)
    
    if not account:
        return jsonify({"msg": "Account not found"}), 404
    
    db.session.delete(account)
    db.session.commit()
    
    return jsonify({"msg": "Account deleted successfully"}), 200


@account_bp.route('/api/account/classrooms', methods=['GET'])
@jwt_required()
def get_account_classrooms():
    """Get all classrooms for the current account"""
    account_id = get_jwt_identity()
    account = Account.query.get(account_id)
    
    if not account:
        return jsonify({"msg": "Account not found"}), 404
    
    classrooms = []
    for classroom in account.classrooms:
        friends_count = classroom.sent_relations.count()
        
        classrooms.append({
            "id": classroom.id,
            "name": classroom.name,
            "location": classroom.location,
            "latitude": classroom.lattitude,
            "longitude": classroom.longitude,
            "class_size": classroom.class_size,
            "availability": classroom.availability,
            "interests": classroom.interests,
            "friends_count": friends_count
        })
    
    return jsonify({"classrooms": classrooms}), 200